keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,845 @@
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+For clients moving from no proxy to proxy, a modification to the kubeconfig
+must be made and a proxy binary must be located as specified in the kubeconfig.
+For clients moving from one version of the proxy to another, the proxy binary
+must be upgraded.
+
+### Version Skew Strategy
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+The proxy binary may be required to be updated before the client is updated in
+situations where the API over stdio between them changes.
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness/README.md.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: ClientExecProxy
+    - Components depending on the feature gate: kubelet, kubectl,
```

Comment:
How will this work for `kubectl` which does not support feature gates AFAIK?

Reply:
What did we decide in SIG Auth? Prefix fields with alpha while this functionality is in the alpha stage?

Reply:
The mechanism has to reflect user choice / opt-in.  The fields being alpha is meaningless.  My preference would be to check an interface which is implemented by feature gates on everything but kubectl.  kubectl would use an environmental variable based implementation.

Reply:
Yes, env var is usually what we do for kubectl to enable alpha flags/features, as I recall. The primary point of gating CLI features is so that the user of those features is sure to be aware that the feature is alpha, and will not accidentally use it thinking it is GA.

Presumably for controllers consuming these libraries, they will have a way to know this is alpha?

Reply:
> Presumably for controllers consuming these libraries, they will have a way to know this is alpha?
I think the feature gate documentation should be sufficient. [1]

[1]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/